Enable running ORT in a Docker container

iamwillbar opened the pull request:
This adds a Dockerfile that allows the ORT scripts to be run inside a container (to avoid having to install dependencies such as a JDK) on the host.

Ideally, the oss-review-toolkit team would build the image and publish it to DockerHub so consumers don't need to clone the repository and build the image themselves. Let me know if you'd like me to help this out.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/heremaps/oss-review-toolkit/141)
<!-- Reviewable:end -->

sschuberth commented:
Thanks for this!

> Ideally, the oss-review-toolkit team would build the image and publish it to DockerHub so consumers don't need to clone the repository and build the image themselves.

We have this actually in our sprint 😉

sschuberth commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM openjdk:8
```
Why openjdk? In the past, I had issues with openjdk from time to time, which were resolved by using Oracle JDK.

sschuberth replied:
Ah, so it's because [we're not allowed to redistribute the Oracle JDK](https://devops.stackexchange.com/a/434).

iamwillbar replied:
The "java" docker image is deprecated and the recommended replacement was "openjdk":
https://hub.docker.com/_/java/

sschuberth commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+# Change to non-root
+RUN groupadd -r toolkit && useradd --no-log-init -r -g toolkit toolkit
+USER toolkit
```
Nit: Please add a final newline.

iamwillbar replied:
Fixed

sschuberth replied:
Hmm, did you actually (force-)push the new version?

iamwillbar replied:
I hit issues testing (moving the -o pipefail didn't work) so I haven't pushed, stay tuned.

sschuberth commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+RUN ["/bin/bash", "-c", "set -o pipefail && curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -"]
+RUN ["/bin/bash", "-c", "set -o pipefail && echo \"deb https://dl.yarnpkg.com/debian/ stable main\" | tee /etc/apt/sources.list.d/yarn.list"]
+RUN apt-get update && apt-get install -y apt-utils 
+RUN apt-get install -y build-essential python-pip nodejs yarn && pip install virtualenv && rm -rf /var/lib/apt/lists
```
I wonder whether we should hard-code the tool versions we currently require by default. That's the way we currently do it in our internal `Dockerfile`:

    ...
    # install Node.js 8.x and NPM 5.3.0 (see: https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions)
        curl -sL https://deb.nodesource.com/setup_8.x | bash - && \
        apt-get install -y nodejs && \
        npm install -g npm@5.5.1 && \
    # install Yarn (see: https://yarnpkg.com/lang/en/docs/install/#linux-tab)
        curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
        echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list && \
        apt-get update && \
        apt-get install -y --no-install-recommends yarn=1.3.2-1 && \
    ...

iamwillbar replied:
Publishing the image will lockdown the versions at the time you build the image but if you rebuilt the image then the versions would change. The trade-off is more management of the Docker image. Let me know what you'd like and I'll gladly make the change.

sschuberth replied:
I tend to prefer the fixed versions to not depend on a "random" version that was current at the time when the image was built, even if that adds more maintenance.

@mnonnenmacher, what's you opinion here?

sschuberth replied:
Also, I believe we should merge this `RUN` line with the one above, so `apt-get update` and `apt-get install` are always run together in the same layer. Otherwise, `apt update` may not be run if we updated only current line 14 [due to layer caching](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#apt-get).

Speaking of `apt-get`, these days I guess we should simply use `apt`.

iamwillbar replied:
Good point, I'll make that change.

mnonnenmacher replied:
@sschuberth I agree that we should use fixed versions of the tools. We only test specific versions and should guarantee that these tested versions are installed in the container.

mnonnenmacher replied:
We also need to install these for the downloader: cvs, subversion, mercurial, git

sschuberth replied:
Correct. My thinking was to only do the obvious fixes for now, then merge, then align the committed with our internal Docker file, publish the container, then change CI to use the container to run tests.

sschuberth commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+RUN ./gradlew installDist
+
+# Install dependencies for NPM scanning
+RUN ["/bin/bash", "-c", "set -o pipefail && curl -sL https://deb.nodesource.com/setup_8.x | bash -"]
```
You might want to pass `-o pipefail` directly to `bash`.

iamwillbar replied:
Thanks, didn't know I could do that.

iamwillbar replied:
Fixed

sschuberth replied:
Also, we're having quite a few `RUN` statements here. AFAIK, each `RUN` [adds a new Docker layer](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#run), and we might want to avoid too many layers, e.g. by putting all `/bin/bash` calls into a single layer.

iamwillbar replied:
I was going to but I wasn't sure I could because of the piping... Relatively new bash user... What would this look like?

sschuberth replied:
I believe in total the line should look like:

    RUN ["/bin/bash", "-o", "pipefail", "-c", "curl -sL https://deb.nodesource.com/setup_8.x | bash - && curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && echo \"deb https://dl.yarnpkg.com/debian/ stable main\" | tee /etc/apt/sources.list.d/yarn.list"]

iamwillbar commented:
OK, I've addressed a lot of the issues except:
* I had issues moving to use -o pipefail instead of the set and it wasn't clear why. The Docker documentation uses the syntax I have here so I reverted back to that.
* I haven't pinned the versions, if you let me know what versions you've tested with I can pin to those or if you provide your internal Dockerfile I can reconcile the differences.

Also, before you publish this to DockerHub you should add some LABEL statements to declare the maintainer etc.

sschuberth commented:
Thanks! Let's just merge this now as-is, I'll improve it later before publishing.

sschuberth commented:
> I had issues moving to use -o pipefail instead of the set and it wasn't clear why.

FYI, I just tried this myself on my Linux box and it worked just fine for me. Note that the order of arguments is important here, `-o` needs to come before `-c`. Maybe that was your issue? Anyway, I'll propose this change as part of an upcoming PR.

sschuberth commented:
@iamwillbar I realize your `README.md` changes you had as part of a previous iteration of this PR did not make it into the final version. Mind filing a new PR only for the `README.md` changes?
